### ISSUE TITLE
could you please describe what ground false means ?

### ISSUE THREAD

**Issue by OhJaeKwang** [thuml/Transfer-Learning-Library#117]
i saw code encoding "ground_false" in  PseudoLabelGenerator2d Class on Keypointdetection
so i want to know what it means and why it was encoded

`
def __init__(self, num_keypoints, height=64, width=64, sigma=2):
        super(PseudoLabelGenerator2d, self).__init__()
        self.height = height
        self.width = width
        self.sigma = sigma

        heatmaps = np.zeros((width, height, height, width), dtype=np.float32)

        tmp_size = sigma * 3
        for mu_x in range(width):
            for mu_y in range(height):
                # Check that any part of the gaussian is in-bounds
                ul = [int(mu_x - tmp_size), int(mu_y - tmp_size)]
                br = [int(mu_x + tmp_size + 1), int(mu_y + tmp_size + 1)]

                # Generate gaussian
                size = 2 * tmp_size + 1
                x = np.arange(0, size, 1, np.float32)
                y = x[:, np.newaxis]
                x0 = y0 = size // 2
                # The gaussian is not normalized, we want the center value to equal 1
                g = np.exp(- ((x - x0) ** 2 + (y - y0) ** 2) / (2 * sigma ** 2))

                # Usable gaussian range
                g_x = max(0, -ul[0]), min(br[0], width) - ul[0]
                g_y = max(0, -ul[1]), min(br[1], height) - ul[1]
                # Image range
                img_x = max(0, ul[0]), min(br[0], width)
                img_y = max(0, ul[1]), min(br[1], height)

                heatmaps[mu_x][mu_y][img_y[0]:img_y[1], img_x[0]:img_x[1]] = \
                    g[g_y[0]:g_y[1], g_x[0]:g_x[1]]

        self.heatmaps = heatmaps
        self.false_matrix = 1. - np.eye(num_keypoints, dtype=np.float32)   

    def forward(self, y):
        B, K, H, W = y.shape   
        y = y.detach()
        preds, max_vals = get_max_preds(y.cpu().numpy())  
        preds = preds.reshape(-1, 2).astype(np.int) 
        ground_truth = self.heatmaps[preds[:, 0], preds[:, 1], :, :].copy().reshape(B, K, H, W).copy()

        ground_false = ground_truth.reshape(B, K, -1).transpose((0, 2, 1)) # B , K , (imgsize) --> B , imgsize , K
        ground_false = ground_false.dot(self.false_matrix).clip(max=1., min=0.).transpose((0, 2, 1)).reshape(B, K, H, W).copy() 
        return torch.from_numpy(ground_truth).to(y.device), torch.from_numpy(ground_false).to(y.device)
`

could you please answer me? Thank you

**Comment by JunguangJiang** [comment]
Since we have no labels on the target domain, we use the predictions to calculate _disparity_ loss.
_ground_truth_ means the predicted correct location of keypoints.
_ground_false_ means the predicted incorrect location of keypoints.

The reason to introduce _ground_false_ can be found in paper [Regressive Domain Adaptation for Unsupervised Keypoint Detection](https://arxiv.org/abs/2103.06175)
